# combine the dataframes
In this notebook we gonna combine the dataframes into one common dataframe, that lists the clusters from the used cluster approaches
The aim is to inspect the clustered data

In [1]:
import pickle
import json
import pandas as pd
import numpy as np


In [2]:
fieldsdf = pd.DataFrame()

with open('obj/onehotdf.pkl', 'rb') as handle:
    fieldsdf = pickle.load(handle)

fieldsdf.head()



swagger  info  host  basePath  schemes  \
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   

                               consumes  produces  paths  definitions  \
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
afterbanks.com                        0         1      1            1   

                               parameters  responses  securityDefinitions  \
1forge.com                              0          0                    0   
6-dot-authentiqio.appspot.com           1          1                    0   
adafruit.com                            1          0                    1   
adobe.com:aem                           0          0                    1   
afterbanks.com                          0          0                    0   

                               security  tags  externalDocs  \
1forge.com                            0     0             0   
6-dot-authentiqio.appspot.com         0     0             0   
adafruit.com                          1     0             0   
adobe.com:aem                         1     0             0   
afterbanks.com                        0     0             0   

                               implemented fields  missing fields  
1forge.com                                      7               8  
6-dot-authentiqio.appspot.com                   9               6  
adafruit.com                                   11               4  
adobe.com:aem                                   9               6  
afterbanks.com                                  8               7

In [3]:
infodf = pd.DataFrame()

with open('obj/infohotdf.pkl', 'rb') as handle:
    infodf = pickle.load(handle)

infodf.head()


title  description  termsOfService  contact  \
1forge.com                         1            1               0        1   
6-dot-authentiqio.appspot.com      1            1               1        1   
adafruit.com                       1            1               0        0   
adobe.com:aem                      1            1               0        1   
afterbanks.com                     1            1               0        0   

                               license  version  used keys  missing keys  
1forge.com                           0        1          4             2  
6-dot-authentiqio.appspot.com        1        1          6             1  
adafruit.com                         0        1          3             3  
adobe.com:aem                        0        1          4             2  
afterbanks.com                       0        1          3             3

In [4]:
keys_field = list(fieldsdf.keys())
keys_field[:-2]

['swagger',
 'info',
 'host',
 'basePath',
 'schemes',
 'consumes',
 'produces',
 'paths',
 'definitions',
 'parameters',
 'responses',
 'securityDefinitions',
 'security',
 'tags',
 'externalDocs']

In [5]:
field_df = pd.DataFrame(fieldsdf[keys_field[:-2]])

In [6]:
field_df.head()

swagger  info  host  basePath  schemes  \
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   

                               consumes  produces  paths  definitions  \
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
afterbanks.com                        0         1      1            1   

                               parameters  responses  securityDefinitions  \
1forge.com                              0          0                    0   
6-dot-authentiqio.appspot.com           1          1                    0   
adafruit.com                            1          0                    1   
adobe.com:aem                           0          0                    1   
afterbanks.com                          0          0                    0   

                               security  tags  externalDocs  
1forge.com                            0     0             0  
6-dot-authentiqio.appspot.com         0     0             0  
adafruit.com                          1     0             0  
adobe.com:aem                         1     0             0  
afterbanks.com                        0     0             0

In [7]:
keys_info = list(infodf.keys())
keys_info[:-2]

['title', 'description', 'termsOfService', 'contact', 'license', 'version']

In [8]:
info_df = pd.DataFrame(infodf[keys_info[:-2]])

In [9]:
info_df.head()

title  description  termsOfService  contact  \
1forge.com                         1            1               0        1   
6-dot-authentiqio.appspot.com      1            1               1        1   
adafruit.com                       1            1               0        0   
adobe.com:aem                      1            1               0        1   
afterbanks.com                     1            1               0        0   

                               license  version  
1forge.com                           0        1  
6-dot-authentiqio.appspot.com        1        1  
adafruit.com                         0        1  
adobe.com:aem                        0        1  
afterbanks.com                       0        1

In [10]:
info_df = pd.DataFrame(info_df)

In [11]:
field_df = pd.DataFrame(field_df)

In [12]:
type(info_df)

pandas.core.frame.DataFrame

In [13]:
type(field_df)

pandas.core.frame.DataFrame

In [14]:
comp_df = pd.concat([field_df, info_df], axis=1)

In [15]:
comp_df.head()

swagger  info  host  basePath  schemes  \
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   

                               consumes  produces  paths  definitions  \
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
afterbanks.com                        0         1      1            1   

                               parameters   ...     securityDefinitions  \
1forge.com                              0   ...                       0   
6-dot-authentiqio.appspot.com           1   ...                       0   
adafruit.com                            1   ...                       1   
adobe.com:aem                           0   ...                       1   
afterbanks.com                          0   ...                       0   

                               security  tags  externalDocs  title  \
1forge.com                            0     0             0      1   
6-dot-authentiqio.appspot.com         0     0             0      1   
adafruit.com                          1     0             0      1   
adobe.com:aem                         1     0             0      1   
afterbanks.com                        0     0             0      1   

                               description  termsOfService  contact  license  \
1forge.com                               1               0        1        0   
6-dot-authentiqio.appspot.com            1               1        1        1   
adafruit.com                             1               0        0        0   
adobe.com:aem                            1               0        1        0   
afterbanks.com                           1               0        0        0   

                               version  
1forge.com                           1  
6-dot-authentiqio.appspot.com        1  
adafruit.com                         1  
adobe.com:aem                        1  
afterbanks.com                       1  

[5 rows x 21 columns]

In [16]:
comp_df.to_csv('comp_df.csv', encoding='utf8')

In [17]:
with open('obj/'+ 'comp_df' + '.pkl', 'wb') as f:
        pickle.dump(comp_df, f, pickle.HIGHEST_PROTOCOL)


# dataframe cluster structure complete

In [18]:
pamcompdf = pd.read_csv('PAMcomplete.csv',index_col='X')
pamcompdf = pamcompdf.iloc[:,1:]
pamcompdf.head()

swagger  info  host  basePath  schemes  \
X                                                                       
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   

                               consumes  produces  paths  definitions  \
X                                                                       
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
afterbanks.com                        0         1      1            1   

                               parameters         ...          security  tags  \
X                                                 ...                           
1forge.com                              0         ...                 0     0   
6-dot-authentiqio.appspot.com           1         ...                 0     0   
adafruit.com                            1         ...                 1     0   
adobe.com:aem                           0         ...                 1     0   
afterbanks.com                          0         ...                 0     0   

                               externalDocs  title  description  \
X                                                                 
1forge.com                                0      1            1   
6-dot-authentiqio.appspot.com             0      1            1   
adafruit.com                              0      1            1   
adobe.com:aem                             0      1            1   
afterbanks.com                            0      1            1   

                               termsOfService  contact  license  version  \
X                                                                          
1forge.com                                  0        1        0        1   
6-dot-authentiqio.appspot.com               1        1        1        1   
adafruit.com                                0        0        0        1   
adobe.com:aem                               0        1        0        1   
afterbanks.com                              0        0        0        1   

                               pam_fit$clustering  
X                                                  
1forge.com                                      1  
6-dot-authentiqio.appspot.com                   1  
adafruit.com                                    2  
adobe.com:aem                                   2  
afterbanks.com                                  1  

[5 rows x 22 columns]

clusterdata[pam_fit$medoids, ]


                                  
823        landregistry.gov.uk:deed

980                   voodoomfg.com

861                    neowsapp.com

140              amazonaws.com:xray

777 googleapis.com:youtubereporting

997           windows.net:graphrbac  
    
 

In [45]:
pamcompdf.loc[pamcompdf['pam_fit$clustering'] == 1].count()

swagger                139
info                   139
host                   139
basePath               139
schemes                139
consumes               139
produces               139
paths                  139
definitions            139
parameters             139
responses              139
securityDefinitions    139
security               139
tags                   139
externalDocs           139
title                  139
description            139
termsOfService         139
contact                139
license                139
version                139
pam_fit$clustering     139
dtype: int64

In [33]:
pamcompdf.loc[pamcompdf['pam_fit$clustering'] == 1]

swagger  info  host  basePath  schemes  \
X                                                                       
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   
agco-ats.com                         1     1     1         0        1   
apache.org:qakka                     1     1     1         0        1   
apigee.net:marketcheck-cars          1     1     1         1        1   
apis.guru                            1     1     1         1        1   
archive.org:search                   1     1     1         1        1   
archive.org:wayback                  1     1     1         1        1   
aucklandmuseum.com                   1     1     1         0        1   
avaza.com                            1     1     1         0        1   
aviationdata.systems                 1     1     1         0        1   
azure.com:servicefabric              1     1     1         0        1   
beanstream.com                       1     1     1         1        1   
betfair.com                          1     1     1         1        1   
bigoven.com                          1     1     1         0        1   
bikewise.org                         1     1     1         1        1   
bluemix.net:containers               1     1     1         1        1   
ccp.is                               1     1     1         1        1   
circuitsandbox.net                   1     1     1         1        1   
cisco.com                            1     1     1         0        1   
clever-cloud.com                     1     1     1         1        1   
cnab-online.herokuapp.com            1     1     1         1        1   
cybertaxonomy.eu                     1     1     1         1        1   
dataatwork.org                       1     1     1         1        1   
deeparteffects.com                   1     1     1         1        1   
deutschebahn.com:fahrplan            1     1     1         1        1   
deutschebahn.com:flinkster           1     1     1         1        1   
deutschebahn.com:stada               1     1     1         1        1   
dracoon.team                         1     1     1         1        1   
...                                ...   ...   ...       ...      ...   
roaring.io                           1     1     1         1        1   
rummblelabs.com                      1     1     1         0        1   
schooldigger.com                     1     1     1         0        1   
selectpdf.com                        1     1     1         1        1   
semantria.com                        1     1     1         1        1   
shipstation.com                      1     1     1         1        1   
slack.com                            1     1     1         1        1   
slideroom.com                        1     1     1         0        1   
smart-me.com                         1     1     1         0        1   
solarvps.com                         1     1     1         1        1   
staging-ecotaco.com                  1     1     1         1        1   
stoplight.io                         1     1     1         1        1   
stripe.com                           1     1     1         0        1   
surrey.ca:open511                    1     1     1         1        1   
surrey.ca:trafficloops               1     1     1         1        1   
swagger.io:generator                 1     1     1         1        1   
swaggerhub.com                       1     1     1         0        1   
thesmsworks.co.uk                    1     1     1         1        1   
twitter.com                          1     1     1         1        1   
tyk.com                              1     1     1         0        1   
uebermaps.com                        1     1     1         1        1   
victorops.com                        1     1     1         1        1   
visagecloud.com 

# dataframe clusters strucutre info

In [19]:
infodf = pd.read_csv('PAMinfo.csv',index_col='X')
infodf = infodf.iloc[:,1:]
infodf.head()

title  description  termsOfService  contact  \
X                                                                            
1forge.com                         1            1               0        1   
6-dot-authentiqio.appspot.com      1            1               1        1   
adafruit.com                       1            1               0        0   
adobe.com:aem                      1            1               0        1   
afterbanks.com                     1            1               0        0   

                               license  version  used.keys  missing.keys  \
X                                                                          
1forge.com                           0        1          4             2   
6-dot-authentiqio.appspot.com        1        1          6             0   
adafruit.com                         0        1          3             3   
adobe.com:aem                        0        1          4             2   
afterbanks.com                       0        1          3             3   

                               pam_fit$clustering  
X                                                  
1forge.com                                      1  
6-dot-authentiqio.appspot.com                   2  
adafruit.com                                    3  
adobe.com:aem                                   1  
afterbanks.com                                  3

clusterdata[pam_fit$medoids, ]
                       
1016            zuora.com

1007            wowza.com

1015      zoomconnect.com

1014              zoom.us

1005 wordassociations.net

In [43]:
infodf.loc[infodf['pam_fit$clustering'] == 1].count()

title                 240
description           240
termsOfService        240
contact               240
license               240
version               240
used.keys             240
missing.keys          240
pam_fit$clustering    240
dtype: int64

In [44]:
infodf.loc[infodf['pam_fit$clustering'] == 1]

title  description  \
X                                                                        
1forge.com                                              1            1   
adobe.com:aem                                           1            1   
agco-ats.com                                            1            0   
archive.org:search                                      1            1   
archive.org:wayback                                     1            1   
avaza.com                                               1            1   
aviationdata.systems                                    1            0   
azure.com:automation-softwareUpdateConfiguration        1            1   
azure.com:automation-softwareUpdateConfiguratio...      1            1   
azure.com:automation-softwareUpdateConfiguratio...      1            1   
azure.com:cognitiveservices-TextAnalytics               1            1   
azure.com:migrate                                       1            1   
bikewise.org                                            1            1   
carbondoomsday.com                                      1            1   
cisco.com                                               1            1   
clever-cloud.com                                        1            1   
clever.com                                              1            1   
clickmeter.com                                          1            1   
core.ac.uk                                              1            1   
dataatwork.org                                          1            1   
deeparteffects.com                                      1            0   
envoice.in                                              1            1   
etmdb.com                                               1            1   
fantasydata.net:cbb-v3-scores                           1            0   
fantasydata.net:cbb-v3-stats                            1            0   
fantasydata.net:cfb-v3-scores                           1            1   
fantasydata.net:golf-v2                                 1            0   
fantasydata.net:mlb-v3-play-by-play                     1            1   
fantasydata.net:mlb-v3-projections                      1            1   
fantasydata.net:mlb-v3-projections-dfsr                 1            1   
...                                                   ...          ...   
selectpdf.com                                           1            1   
semantria.com                                           1            1   
solarvps.com                                            1            1   
spectrocoin.com                                         1            1   
spinitron.com                                           1            1   
stoplight.io                                            1            0   
storecove.com                                           1            1   
swaggerhub.com                                          1            1   
taggun.io                                               1            1   
thesmsworks.co.uk                                       1            1   
transavia.com                                           1            1   
turbinelabs.io                                          1            1   
tvmaze.com                                              1            1   
twinehealth.com                                         1            1   
twitter.com                                             1            0   
tyk.com                                                 1            0   
victorops.com                                           1            1   
vocadb.net                                              1            0   
voodoomfg.com                                           1            1   
waterlinked.com                                         1            1   
weatherbit.io                                           1            1   
whapi.com:accounts                                      

# dataframe clusters path metrics

In [20]:
pathdf = pd.read_csv('PAMpath.csv',index_col='X')
pathdf = pathdf.iloc[:,1:]
pathdf.head()

paths  operations  pam_fit$clustering
X                                                                   
1forge.com                         2           2                   1
6-dot-authentiqio.appspot.com      5          14                   2
adafruit.com                      34          69                   3
adobe.com:aem                     26          30                   4
afterbanks.com                     3           3                   5

clusterdata[pam_fit$medoids, ]

                                 X paths operations
                                 
7          airport-web.appspot.com     1          1

938                  slideroom.com     7         11

964                    traccar.org    34         57

985                 wealthport.com    15         26

473 azure.com:sql-serverDnsAliases     3          5

625                   figshare.com    80        141

In [78]:
pathdf.loc[pathdf['pam_fit$clustering'] == 6]

paths  operations  \
X                                                                   
agco-ats.com                                      109         191   
amazonaws.com:apigateway                           52         172   
amazonaws.com:autoscaling                          52         104   
amazonaws.com:clouddirectory                       60         122   
amazonaws.com:cognito-idp                          95         190   
amazonaws.com:ec2                                 268         536   
amazonaws.com:email                                70         140   
amazonaws.com:gamelift                             65         130   
amazonaws.com:glue                                 78         156   
amazonaws.com:iam                                 122         244   
amazonaws.com:iot                                  75         198   
amazonaws.com:lightsail                            71         142   
amazonaws.com:opsworks                             74         148   
amazonaws.com:rds                                  91         182   
amazonaws.com:redshift                             63         126   
amazonaws.com:s3                                   74         151   
amazonaws.com:servicecatalog                       60         120   
amazonaws.com:ssm                                  97         194   
amazonaws.com:storagegateway                       63         126   
amazonaws.com:waf                                  69         138   
amazonaws.com:waf-regional                         73         146   
apacta.com                                         79         136   
api2cart.com                                      108         108   
azure.com:cognitiveservices-LUIS-Programmatic      56          97   
azure.com:devtestlabs-DTL                          64         107   
azure.com:network                                  63          97   
azure.com:recoveryservicessiterecovery-service     80         104   
azure.com:servicefabric                           102         113   
azure.com:storsimple8000series-storsimple          66          92   
azure.com:web-WebApps                             224         311   
...                                               ...         ...   
googleapis.com:dfareporting                       117         321   
googleapis.com:games                               52         105   
illumidesk.com                                     71         144   
iva-api.com                                       116         116   
jumpseller.com                                     56          96   
kubernetes.io                                     488        1381   
magento.com                                       277         355   
mandrillapp.com                                    90          90   
naviplancentral.com:factfinder                     73         129   
nba.com                                            91         182   
osf.io                                             90         112   
osisoft.com                                       218         345   
payrun.io                                         205         263   
rebilly.com                                        80         197   
reverb.com                                        125         162   
sendgrid.com                                      144         304   
slack.com                                         135         135   
slicebox.local                                     86         105   
squareup.com                                       63          96   
stackexchange.com                                 124         124   
stripe.com                                        115         197   
tfl.gov.uk                                         85          85   
trello.com                                        264         324   
twilio.com                                         63         116   
twitter.com                                        82         165   
vocadb.net                                    

In [74]:
pathdf.loc[pathdf['pam_fit$clustering'] == 2].count()

paths                 178
operations            178
pam_fit$clustering    178
dtype: int64

In [95]:
for i in range (1,7):
    print("#####################")
    print("cluster: "+ str(i))
    print("#####################")
    print("min:" + str(pathdf.loc[pathdf['pam_fit$clustering'] == i].min()))
    print("max:" + str(pathdf.loc[pathdf['pam_fit$clustering'] == i].max()))
   

#####################
cluster: 1
#####################
min:paths                 1
operations            1
pam_fit$clustering    1
dtype: int64
max:paths                 2
operations            2
pam_fit$clustering    1
dtype: int64
#####################
cluster: 2
#####################
min:paths                 3
operations            8
pam_fit$clustering    2
dtype: int64
max:paths                 15
operations            17
pam_fit$clustering     2
dtype: int64
#####################
cluster: 3
#####################
min:paths                 16
operations            34
pam_fit$clustering     3
dtype: int64
max:paths                  68
operations            112
pam_fit$clustering      3
dtype: int64
#####################
cluster: 4
#####################
min:paths                  5
operations            16
pam_fit$clustering     4
dtype: int64
max:paths                 32
operations            43
pam_fit$clustering     4
dtype: int64
#####################
cluster: 5
#################

# dataframe clusters fields  count structure

In [21]:

fieldcompdf = pd.read_csv('PAMcomp.csv',index_col='X')
fieldcompdf = fieldcompdf.iloc[:,1:]
fieldcompdf.head()


swagger  info  host  basePath  schemes  \
X                                                                       
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   

                               consumes  produces  paths  definitions  \
X                                                                       
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
afterbanks.com                        0         1      1            1   

                               parameters  responses  securityDefinitions  \
X                                                                           
1forge.com                              0          0                    0   
6-dot-authentiqio.appspot.com           1          1                    0   
adafruit.com                            1          0                    1   
adobe.com:aem                           0          0                    1   
afterbanks.com                          0          0                    0   

                               security  tags  externalDocs  \
X                                                             
1forge.com                            0     0             0   
6-dot-authentiqio.appspot.com         0     0             0   
adafruit.com                          1     0             0   
adobe.com:aem                         1     0             0   
afterbanks.com                        0     0             0   

                               implemented.fields  missing.fields  cluster  \
X                                                                            
1forge.com                                      7               8        0   
6-dot-authentiqio.appspot.com                   9               6        1   
adafruit.com                                   11               4        2   
adobe.com:aem                                   9               6        2   
afterbanks.com                                  8               7        0   

                               pam_fit$clustering  
X                                                  
1forge.com                                      1  
6-dot-authentiqio.appspot.com                   2  
adafruit.com                                    3  
adobe.com:aem                                   2  
afterbanks.com                                  1

clusterdata[pam_fit$medoids, ]

                      X swagger info host basePath schemes consumes produces paths definitions parameters responses
                      
1011          yunbi.com       1    1    1        1       1        0        1     1           0          0         0

1013        zappiti.com       1    1    1        1       1        1        1     1           1          0         0

1016          zuora.com       1    1    1        1       1        1        1     1           1          1         0

1015    zoomconnect.com       1    1    1        1       1        0        1     1           1          0         0

925  rottentomatoes.com       1    1    1        1       1        1        1     1           1          0         0

In [94]:
fieldcompdf.loc[pathdf['pam_fit$clustering'] == 3].head()

swagger  info  host  basePath  schemes  \
X                                                                        
adafruit.com                          1     1     1         1        1   
amazonaws.com:alexaforbusiness        1     1     1         1        1   
amazonaws.com:appstream               1     1     1         1        1   
amazonaws.com:cloudformation          1     1     1         1        1   
amazonaws.com:cloudfront              1     1     1         1        1   

                                consumes  produces  paths  definitions  \
X                                                                        
adafruit.com                           0         1      1            1   
amazonaws.com:alexaforbusiness         1         1      1            1   
amazonaws.com:appstream                1         1      1            1   
amazonaws.com:cloudformation           1         1      1            1   
amazonaws.com:cloudfront               1         1      1            1   

                                parameters  responses  securityDefinitions  \
X                                                                            
adafruit.com                             1          0                    1   
amazonaws.com:alexaforbusiness           1          0                    1   
amazonaws.com:appstream                  1          0                    1   
amazonaws.com:cloudformation             1          0                    1   
amazonaws.com:cloudfront                 1          0                    1   

                                security  tags  externalDocs  \
X                                                              
adafruit.com                           1     0             0   
amazonaws.com:alexaforbusiness         1     0             1   
amazonaws.com:appstream                1     0             1   
amazonaws.com:cloudformation           1     0             1   
amazonaws.com:cloudfront               1     0             1   

                                implemented.fields  missing.fields  cluster  \
X                                                                             
adafruit.com                                    11               4        2   
amazonaws.com:alexaforbusiness                  13               2        4   
amazonaws.com:appstream                         13               2        4   
amazonaws.com:cloudformation                    13               2        4   
amazonaws.com:cloudfront                        13               2        4   

                                pam_fit$clustering  
X                                                   
adafruit.com                                     3  
amazonaws.com:alexaforbusiness                   5  
amazonaws.com:appstream                          5  
amazonaws.com:cloudformation                     5  
amazonaws.com:cloudfront                         5

In [93]:
fieldstructdf.loc[pathdf['pam_fit$clustering'] == 3].head()

swagger  info  host  basePath  schemes  \
X                                                                        
adafruit.com                          1     1     1         1        1   
amazonaws.com:alexaforbusiness        1     1     1         1        1   
amazonaws.com:appstream               1     1     1         1        1   
amazonaws.com:cloudformation          1     1     1         1        1   
amazonaws.com:cloudfront              1     1     1         1        1   

                                consumes  produces  paths  definitions  \
X                                                                        
adafruit.com                           0         1      1            1   
amazonaws.com:alexaforbusiness         1         1      1            1   
amazonaws.com:appstream                1         1      1            1   
amazonaws.com:cloudformation           1         1      1            1   
amazonaws.com:cloudfront               1         1      1            1   

                                parameters  responses  securityDefinitions  \
X                                                                            
adafruit.com                             1          0                    1   
amazonaws.com:alexaforbusiness           1          0                    1   
amazonaws.com:appstream                  1          0                    1   
amazonaws.com:cloudformation             1          0                    1   
amazonaws.com:cloudfront                 1          0                    1   

                                security  tags  externalDocs  \
X                                                              
adafruit.com                           1     0             0   
amazonaws.com:alexaforbusiness         1     0             1   
amazonaws.com:appstream                1     0             1   
amazonaws.com:cloudformation           1     0             1   
amazonaws.com:cloudfront               1     0             1   

                                pam_fit$clustering  
X                                                   
adafruit.com                                     2  
amazonaws.com:alexaforbusiness                   3  
amazonaws.com:appstream                          3  
amazonaws.com:cloudformation                     3  
amazonaws.com:cloudfront                         3

In [83]:
for i in range(1,6):
    print("#########")
    print("cluster" + str(i))
    print("#########")
    print(str(fieldcompdf.loc[pathdf['pam_fit$clustering'] == i].count()))


#########
cluster1
#########
swagger                193
info                   193
host                   193
basePath               193
schemes                193
consumes               193
produces               193
paths                  193
definitions            193
parameters             193
responses              193
securityDefinitions    193
security               193
tags                   193
externalDocs           193
implemented.fields     193
missing.fields         193
cluster                193
pam_fit$clustering     193
dtype: int64
#########
cluster2
#########
swagger                178
info                   178
host                   178
basePath               178
schemes                178
consumes               178
produces               178
paths                  178
definitions            178
parameters             178
responses              178
securityDefinitions    178
security               178
tags                   178
externalDocs           178
implemented

In [99]:
for i in range (1,6):
    print("#####################")
    print("cluster: "+ str(i))
    print("#####################")
    print("min:" + str(fieldcompdf.loc[pathdf['pam_fit$clustering'] == i].min()))
    print("##")
    print("max:" + str(fieldcompdf.loc[pathdf['pam_fit$clustering'] == i].max()))
   

#####################
cluster: 1
#####################
min:swagger                1
info                   1
host                   1
basePath               0
schemes                1
consumes               0
produces               0
paths                  1
definitions            0
parameters             0
responses              0
securityDefinitions    0
security               0
tags                   0
externalDocs           0
implemented.fields     7
missing.fields         2
cluster                0
pam_fit$clustering     1
dtype: int64
##
max:swagger                 1
info                    1
host                    1
basePath                1
schemes                 1
consumes                1
produces                1
paths                   1
definitions             1
parameters              1
responses               1
securityDefinitions     1
security                1
tags                    1
externalDocs            1
implemented.fields     13
missing.fields          8
clus

# dataframe fields structure clusters

In [22]:
fieldstructdf = pd.read_csv('PAMStruct.csv',index_col='X')
fieldstructdf = fieldstructdf.iloc[:,1:]
fieldstructdf.head()

swagger  info  host  basePath  schemes  \
X                                                                       
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   

                               consumes  produces  paths  definitions  \
X                                                                       
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
afterbanks.com                        0         1      1            1   

                               parameters  responses  securityDefinitions  \
X                                                                           
1forge.com                              0          0                    0   
6-dot-authentiqio.appspot.com           1          1                    0   
adafruit.com                            1          0                    1   
adobe.com:aem                           0          0                    1   
afterbanks.com                          0          0                    0   

                               security  tags  externalDocs  \
X                                                             
1forge.com                            0     0             0   
6-dot-authentiqio.appspot.com         0     0             0   
adafruit.com                          1     0             0   
adobe.com:aem                         1     0             0   
afterbanks.com                        0     0             0   

                               pam_fit$clustering  
X                                                  
1forge.com                                      1  
6-dot-authentiqio.appspot.com                   1  
adafruit.com                                    2  
adobe.com:aem                                   1  
afterbanks.com                                  1

clusterdata[pam_fit$medoids, ]

                                  X swagger info host basePath schemes consumes produces paths definitions parameters
                                  
198                       avaza.com       1    1    1        0       1        0        1     1           1          0

997           windows.net:graphrbac       1    1    1        0       1        1        1     1           1          1

516                      bbci.co.uk       1    1    1        1       1        1        1     1           1          1

148  apitore.com:2chMatomeFeedsApis       1    1    1        1       1        1        1     1           1          0

777 googleapis.com:youtubereporting       1    1    1        1       1        0        0     1           1          1

In [88]:
fieldstructdf.loc[pathdf['pam_fit$clustering'] == 1].head()

swagger  info  host  basePath  schemes  \
X                                                                               
1forge.com                                   1     1     1         1        1   
airport-web.appspot.com                      1     1     1         1        1   
amazonaws.com:entitlement.marketplace        1     1     1         1        1   
amazonaws.com:kinesis-video-media            1     1     1         1        1   
amazonaws.com:mobileanalytics                1     1     1         1        1   

                                       consumes  produces  paths  definitions  \
X                                                                               
1forge.com                                    0         1      1            0   
airport-web.appspot.com                       1         1      1            1   
amazonaws.com:entitlement.marketplace         1         1      1            1   
amazonaws.com:kinesis-video-media             1         1      1            1   
amazonaws.com:mobileanalytics                 1         1      1            1   

                                       parameters  responses  \
X                                                              
1forge.com                                      0          0   
airport-web.appspot.com                         0          0   
amazonaws.com:entitlement.marketplace           1          0   
amazonaws.com:kinesis-video-media               1          0   
amazonaws.com:mobileanalytics                   1          0   

                                       securityDefinitions  security  tags  \
X                                                                            
1forge.com                                               0         0     0   
airport-web.appspot.com                                  1         0     0   
amazonaws.com:entitlement.marketplace                    1         1     0   
amazonaws.com:kinesis-video-media                        1         1     0   
amazonaws.com:mobileanalytics                            1         1     0   

                                       externalDocs  pam_fit$clustering  
X                                                                        
1forge.com                                        0                   1  
airport-web.appspot.com                           0                   1  
amazonaws.com:entitlement.marketplace             1                   3  
amazonaws.com:kinesis-video-media                 1                   3  
amazonaws.com:mobileanalytics                     1                   3

In [81]:
for i in range(1,6):
    print("#########")
    print("cluster" + str(i))
    print("#########")
    print(str(fieldstructdf.loc[pathdf['pam_fit$clustering'] == i].count()))


#########
cluster1
#########
swagger                193
info                   193
host                   193
basePath               193
schemes                193
consumes               193
produces               193
paths                  193
definitions            193
parameters             193
responses              193
securityDefinitions    193
security               193
tags                   193
externalDocs           193
pam_fit$clustering     193
dtype: int64
#########
cluster2
#########
swagger                178
info                   178
host                   178
basePath               178
schemes                178
consumes               178
produces               178
paths                  178
definitions            178
parameters             178
responses              178
securityDefinitions    178
security               178
tags                   178
externalDocs           178
pam_fit$clustering     178
dtype: int64
#########
cluster3
#########
swagger                

In [100]:
for i in range (1,6):
    print("#####################")
    print("cluster: "+ str(i))
    print("#####################")
    print("min:" + str(fieldstructdf.loc[pathdf['pam_fit$clustering'] == i].min()))
    print("##")
    print("max:" + str(fieldstructdf.loc[pathdf['pam_fit$clustering'] == i].max()))
   

#####################
cluster: 1
#####################
min:swagger                1
info                   1
host                   1
basePath               0
schemes                1
consumes               0
produces               0
paths                  1
definitions            0
parameters             0
responses              0
securityDefinitions    0
security               0
tags                   0
externalDocs           0
pam_fit$clustering     1
dtype: int64
##
max:swagger                1
info                   1
host                   1
basePath               1
schemes                1
consumes               1
produces               1
paths                  1
definitions            1
parameters             1
responses              1
securityDefinitions    1
security               1
tags                   1
externalDocs           1
pam_fit$clustering     5
dtype: int64
#####################
cluster: 2
#####################
min:swagger                1
info                   1